File: src/ingestion/src/domain/wire.rs
//! C01 Rust boundary validation uses generated descriptors and common field-rule extensions.
//! Bounded decoding preserves unknown fields; validation never opens models/storage or resolves remote IDs.
//! Shared scalar/temporal/vector checks protect worker inputs. Production graph/storage invariants belong
//! to their domain owners; performance gates require real workloads, not these compatibility fixtures.
use crate::wire::{common, documents, evidence, graph, inference};
use protobuf::reflect::ReflectValueRef;
use protobuf::{CodedInputStream, MessageDyn};
use std::collections::HashSet;

#[derive(Clone, Copy)]
pub struct Limits {
    pub max_bytes: usize,
    pub max_depth: u32,
    pub max_items: usize,
}
impl Default for Limits {
    fn default() -> Self {
        Self {
            max_bytes: 16 << 20,
            max_depth: 64,
            max_items: 100000,
        }
    }
}
fn require(ok: bool, field: &str) -> Result<(), String> {
    if ok {
        Ok(())
    } else {
        Err(field.to_string())
    }
}

pub fn validate(message: &dyn MessageDyn, limits: Limits) -> Result<(), String> {
    require(
        limits.max_bytes > 0 && limits.max_depth > 0 && limits.max_items > 0,
        "invalid limits",
    )?;
    require(
        message.compute_size_dyn() <= limits.max_bytes as u64,
        "byte limit",
    )?;
    walk(message, 0, limits, &mut limits.max_items.clone(), "")
}
fn walk(
    m: &dyn MessageDyn,
    depth: u32,
    limits: Limits,
    remaining: &mut usize,
    inherited: &str,
) -> Result<(), String> {
    let own = corpus(m);
    require(
        own.is_empty() || inherited.is_empty() || own == inherited,
        "nested corpus mismatch",
    )?;
    let active = if own.is_empty() { inherited } else { &own };
    require(depth <= limits.max_depth, "depth limit")?;
    let descriptor = m.descriptor_dyn();
    for f in descriptor.fields() {
        let rule = f
            .proto()
            .options
            .as_ref()
            .and_then(|o| common::exts::rules.get(o))
            .unwrap_or_default();
        let where_ = f.full_name();
        let repeated = f.is_repeated();
        require(
            !rule.required
                || if repeated {
                    !f.get_repeated(m).is_empty()
                } else {
                    f.has_field(m)
                },
            &where_,
        )?;
        let values: Vec<ReflectValueRef> = if repeated {
            let list = f.get_repeated(m);
            require(list.len() >= rule.min_items as usize, &where_)?;
            require(list.len() <= *remaining, "item limit")?;
            *remaining -= list.len();
            (0..list.len()).map(|i| list.get(i)).collect()
        } else if !f.has_field(m)
            && (f.proto().proto3_optional()
                || f.containing_oneof_including_synthetic().is_some()
                || matches!(
                    f.singular_runtime_type(),
                    protobuf::reflect::RuntimeType::Message(_)
                ))
        {
            Vec::new()
        } else {
            vec![f.get_singular_field_or_default(m)]
        };
        let mut seen = HashSet::new();
        for v in values {
            if rule.unique {
                require(seen.insert(format!("{v:?}")), &where_)?;
            }
            if let ReflectValueRef::Message(ref child) = v {
                require(*remaining > 0, "item limit")?;
                *remaining -= 1;
                walk(&**child, depth + 1, limits, remaining, active)?;
                continue;
            }
            if let ReflectValueRef::String(s) = v {
                if rule.ascii_id {
                    require(
                        !s.is_empty()
                            && s.len() <= 256
                            && s.bytes().all(|c| (33..=126).contains(&c)),
                        &where_,
                    )?;
                }
                if rule.sha256 {
                    require(
                        s.len() == 64
                            && s.bytes()
                                .all(|c| c.is_ascii_digit() || (b'a'..=b'f').contains(&c)),
                        &where_,
                    )?;
                }
            }
            if let ReflectValueRef::Enum(ref en, n) = v {
                if rule.required {
                    require(n != 0 && en.value_by_number(n).is_some(), &where_)?;
                }
            }
            let num = match v {
                ReflectValueRef::I32(n) => Some(n as f64),
                ReflectValueRef::I64(n) => Some(n as f64),
                ReflectValueRef::U32(n) => Some(n as f64),
                ReflectValueRef::U64(n) => Some(n as f64),
                ReflectValueRef::F32(n) => Some(n as f64),
                ReflectValueRef::F64(n) => Some(n),
                _ => None,
            };
            if let Some(n) = num {
                if rule.positive || rule.finite || rule.probability {
                    require(
                        n.is_finite()
                            && (!rule.positive || n > 0.)
                            && (!rule.probability || (0. ..=1.).contains(&n)),
                        &where_,
                    )?;
                }
                if f.name() == "schema_version" {
                    require(n == 1., "schema version")?;
                }
            }
        }
    }
    for oneof in descriptor.oneofs() {
        require(oneof.fields().any(|f| f.has_field(m)), &oneof.full_name())?;
    }
    semantic(m)
}
fn corpus(m: &dyn MessageDyn) -> String {
    let d = m.descriptor_dyn();
    if let Some(f) = d.field_by_name("corpus_id") {
        return f
            .get_singular_field_or_default(m)
            .to_str()
            .unwrap_or("")
            .to_string();
    }
    for name in ["meta", "context", "batch"] {
        if let Some(f) = d.field_by_name(name) {
            if f.has_field(m) {
                if let Some(child) = f.get_singular(m).and_then(|v| v.to_message()) {
                    let c = corpus(&*child);
                    if !c.is_empty() {
                        return c;
                    }
                }
            }
        }
    }
    String::new()
}
fn semantic(m: &dyn MessageDyn) -> Result<(), String> {
    if let Some(p) = m.downcast_ref::<common::ArtifactRef>() {
        let k = &p.storage_key;
        require(
            !k.contains(':')
                && !k.contains('\\')
                && k.split('/')
                    .all(|part| !part.is_empty() && part != "." && part != ".."),
            "ArtifactRef storage key",
        )?;
    }
    if let Some(p) = m.downcast_ref::<common::TemporalScope>() {
        require(
            p.mode.value() != common::TemporalMode::TEMPORAL_MODE_AS_OF as i32
                || p.effective_at.is_some(),
            "TemporalScope AS_OF",
        )?;
        require(
            if p.mode.value() == common::TemporalMode::TEMPORAL_MODE_COMPARE as i32 {
                p.compare_dates.len() >= 2 && p.effective_at.is_none()
            } else {
                p.compare_dates.is_empty()
            },
            "TemporalScope compare",
        )?;
    }
    if let Some(p) = m.downcast_ref::<graph::GraphPath>() {
        require(
            p.ordered_node_ids.len() == p.ordered_assertion_ids.len() + 1
                && p.selected_support_ids.len() == p.ordered_assertion_ids.len(),
            "GraphPath cardinality",
        )?;
    }
    if let Some(p) = m.downcast_ref::<documents::SourceBlob>() {
        require(
            p.byte_size == p.artifact_ref.byte_size
                && p.raw_sha256.sha256 == p.artifact_ref.content_hash.sha256,
            "SourceBlob artifact",
        )?;
    }
    if let Some(p) = m.downcast_ref::<documents::SourceObservation>() {
        require(
            p.status.value() != documents::ObservationStatus::OBSERVATION_STATUS_COMPLETE as i32
                || (p.source_blob_id.is_some() && p.error.is_none()),
            "SourceObservation completion",
        )?;
    }

    if let Some(p) = m.downcast_ref::<common::CalendarDate>() {
        let y = p.year;
        require(
            (1..=9999).contains(&y) && (1..=12).contains(&p.month),
            "CalendarDate",
        )?;
        let mut days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
        if y % 4 == 0 && (y % 100 != 0 || y % 400 == 0) {
            days[1] = 29
        };
        require(
            p.day > 0 && p.day <= days[p.month as usize - 1],
            "CalendarDate",
        )?;
    }
    if let Some(p) = m.downcast_ref::<common::DateAssertion>() {
        require(
            (p.knowledge.value() == common::DateKnowledge::DATE_KNOWLEDGE_KNOWN as i32)
                == p.value.is_some(),
            "DateAssertion value",
        )?;
        require(
            if p.knowledge.value() == common::DateKnowledge::DATE_KNOWLEDGE_CONFLICT as i32 {
                p.alternatives.len() >= 2
            } else {
                p.alternatives.is_empty()
            },
            "DateAssertion alternatives",
        )?;
    }
    if let Some(p) = m.downcast_ref::<common::Visibility>() {
        require(p.to_seq.is_none_or(|v| v > p.from_seq), "Visibility")?;
    }
    if let Some(p) = m.downcast_ref::<common::TextSpan>() {
        require(p.end_byte >= p.start_byte, "TextSpan")?;
    }
    if let Some(p) = m.downcast_ref::<common::BoundingBox>() {
        require(p.x1 >= p.x0 && p.y1 >= p.y0, "BoundingBox")?;
    }
    if let Some(p) = m.downcast_ref::<common::LegalInterval>() {
        if let (Some(s), Some(e)) = (p.start.value.as_ref(), p.end.value.as_ref()) {
            let number = |d: &common::CalendarDate| {
                i64::from(d.year) * 10000 + i64::from(d.month) * 100 + i64::from(d.day)
            };
            require(number(s) < number(e), "LegalInterval")?;
        }
    }
    if let Some(p) = m.downcast_ref::<evidence::SparseVector>() {
        require(
            p.indices.len() == p.values.len() && p.indices.windows(2).all(|v| v[0] < v[1]),
            "SparseVector",
        )?;
    }
    if let Some(p) = m.downcast_ref::<evidence::DenseVector>() {
        require(p.dimensions as usize == p.values.len(), "DenseVector")?;
    }
    if let Some(p) = m.downcast_ref::<common::Counts>() {
        require(
            p.accepted <= p.expected && p.rejected == p.expected - p.accepted,
            "Counts",
        )?;
    }
    if let Some(p) = m.downcast_ref::<common::TruncationInfo>() {
        require(
            p.retained_tokens <= p.original_tokens
                && (p.truncated || p.retained_tokens == p.original_tokens),
            "TruncationInfo",
        )?;
    }
    if let Some(p) = m.downcast_ref::<common::RequestContext>() {
        if let Some(s) = p.snapshot_ref.as_ref() {
            require(p.corpus_id == s.corpus_id, "context corpus")?;
        }
    }
    if let Some(p) = m.downcast_ref::<protobuf::well_known_types::timestamp::Timestamp>() {
        require(
            (-62135596800..=253402300799).contains(&p.seconds)
                && (0..=999999999).contains(&p.nanos),
            "Timestamp",
        )?;
    }
    if let Some(p) = m.downcast_ref::<common::ModelManifest>() {
        require(
            p.task.value() != common::ModelTask::MODEL_TASK_EMBED as i32 || p.dimensions.is_some(),
            "model dimensions",
        )?;
    }
    if let Some(p) = m.downcast_ref::<inference::EmbedBatchRequest>() {
        let ids: HashSet<_> = p.items.iter().map(|v| &v.item_id).collect();
        require(
            ids.len() == p.items.len()
                && p.model.task.value() == common::ModelTask::MODEL_TASK_EMBED as i32,
            "EmbedBatch",
        )?;
    }
    if let Some(p) = m.downcast_ref::<inference::RerankBatchRequest>() {
        let ids: HashSet<_> = p.pairs.iter().map(|v| &v.pair_id).collect();
        require(
            ids.len() == p.pairs.len()
                && p.model.task.value() == common::ModelTask::MODEL_TASK_RERANK as i32,
            "RerankBatch",
        )?;
    }
    Ok(())
}
pub fn check_utf8_span(text: &[u8], start: u64, end: u64) -> Result<(), String> {
    let text = std::str::from_utf8(text).map_err(|_| "invalid UTF-8")?;
    require(start <= end && end <= text.len() as u64, "span bounds")?;
    require(
        text.is_char_boundary(start as usize) && text.is_char_boundary(end as usize),
        "UTF-8 split",
    )
}
pub fn decode(
    raw: &[u8],
    descriptor: &protobuf::reflect::MessageDescriptor,
    limits: Limits,
) -> Result<Box<dyn MessageDyn>, String> {
    require(raw.len() <= limits.max_bytes, "byte limit")?;
    let mut input = CodedInputStream::from_bytes(raw);
    input.set_recursion_limit(limits.max_depth);
    let m = descriptor
        .parse_from(&mut input)
        .map_err(|e| e.to_string())?;
    input.check_eof().map_err(|e| e.to_string())?;
    validate(&*m, limits)?;
    Ok(m)
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn unicode_boundaries() {
        let text = "Pasal \u{00e9} \u{2014}".as_bytes();
        assert!(check_utf8_span(text, 0, 7).is_err());
        assert!(check_utf8_span(text, 0, 8).is_ok());
    }
    #[test]
    #[ignore = "requires fixture generation via tests/integration/wire_roundtrip.py"]
    fn wire_fixture_interop() {
        let dir = std::env::var_os("REGULAGRAPH_WIRE_FIXTURES").expect("missing wire fixtures");
        let dir = std::path::PathBuf::from(dir);
        let files = [
            crate::wire::common::file_descriptor(),
            crate::wire::documents::file_descriptor(),
            crate::wire::graph::file_descriptor(),
            crate::wire::evidence::file_descriptor(),
            crate::wire::answers::file_descriptor(),
            crate::wire::jobs::file_descriptor(),
            crate::wire::inference::file_descriptor(),
            crate::wire::evaluation::file_descriptor(),
        ];
        let out = dir.join("rust");
        std::fs::create_dir_all(&out).unwrap();
        for line in std::fs::read_to_string(dir.join("cases.tsv"))
            .unwrap()
            .lines()
        {
            let f: Vec<_> = line.split('\t').collect();
            let desc = files
                .iter()
                .find_map(|d| d.message_by_full_name(&format!(".{}", f[1])))
                .unwrap();
            let raw = std::fs::read(dir.join(format!("{}.bin", f[0]))).unwrap();
            let limits = Limits {
                max_bytes: f[3].parse().unwrap(),
                ..Limits::default()
            };
            let result = decode(&raw, &desc, limits);
            assert_eq!(
                result.is_ok(),
                f[2] == "valid",
                "{}: {:?}",
                f[0],
                result.err()
            );
            let m = desc.parse_from_bytes(&raw).unwrap();
            std::fs::write(
                out.join(format!("{}.bin", f[0])),
                m.write_to_bytes_dyn().unwrap(),
            )
            .unwrap();
        }
    }
}

File: src/server/internal/domain/wire.go
// C01 boundary validation consumes generated Protobuf descriptors and shared field rules.
// Input is a bounded decoded message; errors identify invariant/field without dumping content.
// Cross-record checks supplement scalar rules; external registry/storage checks remain S01 owners.
// Limits protect memory/recursion; benchmark targets remain configs/benchmark-targets.yaml, unmeasured.
package domain

import (
	"bytes"
	"errors"
	"fmt"
	"math"
	"net/url"
	"path"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/reflect/protoreflect"
	"google.golang.org/protobuf/types/descriptorpb"
	"google.golang.org/protobuf/types/known/timestamppb"
	pb "regulagraph.local/server/gen/regulagraph/v1"
)

type WireLimits struct {
	MaxBytes int
	MaxDepth int
	MaxItems int
}

var DefaultWireLimits = WireLimits{MaxBytes: 16 << 20, MaxDepth: 64, MaxItems: 100000}

func DecodeWire(raw []byte, dst proto.Message, limits WireLimits) error {
	if limits.MaxBytes <= 0 || limits.MaxDepth <= 0 || limits.MaxItems <= 0 {
		return errors.New("invalid wire limits")
	}
	if len(raw) > limits.MaxBytes {
		return errors.New("wire payload exceeds byte limit")
	}
	if err := (proto.UnmarshalOptions{DiscardUnknown: false, RecursionLimit: limits.MaxDepth}).Unmarshal(raw, dst); err != nil {
		return err
	}
	return ValidateWire(dst, limits)
}
func ValidateWire(message proto.Message, limits WireLimits) error {
	if message == nil || !message.ProtoReflect().IsValid() {
		return errors.New("nil wire message")
	}
	if limits.MaxBytes <= 0 || limits.MaxDepth <= 0 || limits.MaxItems <= 0 {
		return errors.New("invalid wire limits")
	}
	if proto.Size(message) > limits.MaxBytes {
		return errors.New("wire payload exceeds byte limit")
	}
	remaining := limits.MaxItems
	var walk func(protoreflect.Message, int, string) error
	walk = func(m protoreflect.Message, depth int, corpus string) error {
		own := wireCorpus(m)
		if own != "" {
			if corpus != "" && corpus != own {
				return errors.New("nested corpus mismatch")
			}
			corpus = own
		}
		if depth > limits.MaxDepth {
			return errors.New("wire nesting exceeds depth limit")
		}
		fields := m.Descriptor().Fields()
		for i := 0; i < fields.Len(); i++ {
			f := fields.Get(i)
			opts := f.Options().(*descriptorpb.FieldOptions)
			r := proto.GetExtension(opts, pb.E_Rules).(*pb.FieldRules)
			if r == nil {
				r = &pb.FieldRules{}
			}
			where := string(f.FullName())
			v := m.Get(f)
			if r.Required && !m.Has(f) {
				return fmt.Errorf("%s: required", where)
			}
			values := []protoreflect.Value{}
			if f.IsList() {
				list := v.List()
				remaining -= list.Len()
				if remaining < 0 {
					return errors.New("wire item limit exceeded")
				}
				if list.Len() < int(r.MinItems) {
					return fmt.Errorf("%s: too few items", where)
				}
				for j := 0; j < list.Len(); j++ {
					values = append(values, list.Get(j))
				}
			} else if !f.HasPresence() || m.Has(f) {
				values = append(values, v)
			}
			seen := map[string]bool{}
			for _, x := range values {
				if r.Unique {
					key := fmt.Sprint(x.Interface())
					if seen[key] {
						return fmt.Errorf("%s: duplicate value", where)
					}
					seen[key] = true
				}
				if f.Kind() == protoreflect.MessageKind {
					remaining--
					if remaining < 0 {
						return errors.New("wire item limit exceeded")
					}
					if err := walk(x.Message(), depth+1, corpus); err != nil {
						return err
					}
					continue
				}
				if f.Kind() == protoreflect.StringKind {
					s := x.String()
					if !utf8.ValidString(s) {
						return fmt.Errorf("%s: invalid UTF-8", where)
					}
					if r.AsciiId && (len(s) == 0 || len(s) > 256 || strings.IndexFunc(s, func(c rune) bool { return c < 33 || c > 126 }) >= 0) {
						return fmt.Errorf("%s: invalid opaque ASCII ID", where)
					}
					if r.Sha256 && (len(s) != 64 || strings.IndexFunc(s, func(c rune) bool { return !(c >= '0' && c <= '9' || c >= 'a' && c <= 'f') }) >= 0) {
						return fmt.Errorf("%s: invalid SHA-256", where)
					}
				}
				if f.Kind() == protoreflect.EnumKind && r.Required && (x.Enum() == 0 || f.Enum().Values().ByNumber(x.Enum()) == nil) {
					return fmt.Errorf("%s: invalid required enum", where)
				}
				if r.Positive || r.Finite || r.Probability {
					var n float64
					switch f.Kind() {
					case protoreflect.FloatKind, protoreflect.DoubleKind:
						n = x.Float()
					case protoreflect.Uint32Kind, protoreflect.Uint64Kind, protoreflect.Fixed32Kind, protoreflect.Fixed64Kind:
						n = float64(x.Uint())
					default:
						n = float64(x.Int())
					}
					if math.IsNaN(n) || math.IsInf(n, 0) || (r.Positive && n <= 0) || (r.Probability && (n < 0 || n > 1)) {
						return fmt.Errorf("%s: numeric constraint", where)
					}
				}
				if f.Name() == "schema_version" && x.Uint() != 1 {
					return fmt.Errorf("%s: unsupported schema version", where)
				}
			}
		}
		oneofs := m.Descriptor().Oneofs()
		for i := 0; i < oneofs.Len(); i++ {
			o := oneofs.Get(i)
			if !o.IsSynthetic() && m.WhichOneof(o) == nil {
				return fmt.Errorf("%s: oneof required", o.FullName())
			}
		}
		return validateSemantics(m.Interface())
	}
	return walk(message.ProtoReflect(), 0, "")
}

func dateNumber(d *pb.CalendarDate) int64 {
	return int64(d.Year)*10000 + int64(d.Month)*100 + int64(d.Day)
}
func validateSemantics(message proto.Message) error {
	bad := func(s string) error { return fmt.Errorf("%s: %s", message.ProtoReflect().Descriptor().Name(), s) }
	switch m := message.(type) {
	case *timestamppb.Timestamp:
		if m.CheckValid() != nil {
			return bad("invalid timestamp")
		}
	case *pb.CalendarDate:
		if m.Year > 9999 || m.Month > 12 || m.Day > 31 {
			return bad("invalid calendar date")
		}
		d := time.Date(int(m.Year), time.Month(m.Month), int(m.Day), 0, 0, 0, 0, time.UTC)
		if d.Year() != int(m.Year) || d.Month() != time.Month(m.Month) || d.Day() != int(m.Day) {
			return bad("invalid calendar date")
		}
	case *pb.DateAssertion:
		if m.Knowledge == pb.DateKnowledge_DATE_KNOWLEDGE_KNOWN {
			if m.Value == nil || len(m.Alternatives) > 0 {
				return bad("known date requires one value")
			}
		} else if m.Value != nil {
			return bad("unknown/conflict/unbounded cannot have a resolved value")
		}
		if m.Knowledge == pb.DateKnowledge_DATE_KNOWLEDGE_CONFLICT {
			if len(m.Alternatives) < 2 {
				return bad("conflict requires alternatives")
			}
		} else if len(m.Alternatives) > 0 {
			return bad("alternatives require conflict status")
		}
	case *pb.LegalInterval:
		if m.Start.Value != nil && m.End.Value != nil && dateNumber(m.Start.Value) >= dateNumber(m.End.Value) {
			return bad("empty or reversed legal interval")
		}
	case *pb.Visibility:
		if m.ToSeq != nil && *m.ToSeq <= m.FromSeq {
			return bad("invalid visibility interval")
		}
	case *pb.TextSpan:
		if m.EndByte < m.StartByte {
			return bad("reversed byte span")
		}
	case *pb.AnswerTextSpan:
		if m.EndByte < m.StartByte {
			return bad("reversed answer span")
		}
	case *pb.BoundingBox:
		if m.X0 > m.X1 || m.Y0 > m.Y1 {
			return bad("reversed bounding box")
		}
	case *pb.ArtifactRef:
		if strings.Contains(m.StorageKey, "\\") || strings.HasPrefix(m.StorageKey, "/") || path.Clean(m.StorageKey) != m.StorageKey || strings.HasPrefix(m.StorageKey, "../") || (m.StorageKey == ".." || m.StorageKey == ".") || strings.Contains(m.StorageKey, ":") {
			return bad("storage key must be a relative artifact key")
		}
	case *pb.RequestContext:
		if m.Deadline.CheckValid() != nil {
			return bad("invalid deadline timestamp")
		}
		if m.SnapshotRef != nil && m.CorpusId != m.SnapshotRef.CorpusId {
			return bad("context/snapshot corpus mismatch")
		}
	case *pb.SourceObservation:
		if m.FetchedAt.CheckValid() != nil {
			return bad("invalid observation timestamp")
		}
		if m.Status == pb.ObservationStatus_OBSERVATION_STATUS_COMPLETE && (m.SourceBlobId == nil || m.Error != nil) {
			return bad("complete observation requires blob and no error")
		}
	case *pb.SourceBlob:
		if m.ByteSize != m.ArtifactRef.ByteSize || m.RawSha256.Sha256 != m.ArtifactRef.ContentHash.Sha256 {
			return bad("blob/artifact mismatch")
		}
	case *pb.TemporalScope:
		if m.Mode == pb.TemporalMode_TEMPORAL_MODE_AS_OF && m.EffectiveAt == nil {
			return bad("AS_OF requires effective date")
		}
		if m.Mode == pb.TemporalMode_TEMPORAL_MODE_COMPARE {
			if len(m.CompareDates) < 2 || m.EffectiveAt != nil {
				return bad("COMPARE requires date list")
			}
		} else if len(m.CompareDates) > 0 {
			return bad("compare dates outside COMPARE")
		}
	case *pb.DenseVector:
		if len(m.Values) != int(m.Dimensions) {
			return bad("dense dimension mismatch")
		}
	case *pb.SparseVector:
		if len(m.Indices) != len(m.Values) {
			return bad("sparse length mismatch")
		}
		for i := 1; i < len(m.Indices); i++ {
			if m.Indices[i] <= m.Indices[i-1] {
				return bad("sparse indices not strictly sorted")
			}
		}
	case *pb.Counts:
		if m.Accepted > m.Expected || m.Rejected != m.Expected-m.Accepted {
			return bad("counts do not balance")
		}
	case *pb.TruncationInfo:
		if m.RetainedTokens > m.OriginalTokens || (!m.Truncated && m.RetainedTokens != m.OriginalTokens) {
			return bad("inconsistent truncation")
		}
	case *pb.GraphPath:
		if len(m.OrderedNodeIds) != len(m.OrderedAssertionIds)+1 || len(m.SelectedSupportIds) != len(m.OrderedAssertionIds) {
			return bad("path edge/node/support cardinality mismatch")
		}
	case *pb.EvidenceBundle:
		ids := map[string]bool{}
		for _, v := range m.Items {
			if ids[v.Meta.RecordId] {
				return bad("duplicate evidence ID")
			}
			ids[v.Meta.RecordId] = true
			if !proto.Equal(v.SnapshotRef, m.Snapshot) || v.Meta.CorpusId != m.Meta.CorpusId {
				return bad("mixed evidence snapshots/corpora")
			}
		}
		if m.Completeness == pb.Completeness_COMPLETENESS_COMPLETE && (len(m.MissingDependencies) > 0 || m.CompletionStatus != pb.CompletionStatus_COMPLETION_STATUS_SUCCEEDED) {
			return bad("incomplete evidence marked complete")
		}
	case *pb.ContextBundle:
		if len(m.OrderedEvidenceIds) != len(m.RenderedBlocks) {
			return bad("context block mismatch")
		}
		for i, b := range m.RenderedBlocks {
			if b.EvidenceId != m.OrderedEvidenceIds[i] {
				return bad("context order mismatch")
			}
		}
		if m.Completeness == pb.Completeness_COMPLETENESS_COMPLETE && len(m.OmittedRequiredRefs) > 0 {
			return bad("omitted evidence marked complete")
		}
	case *pb.Citation:
		if m.SourceSpan == nil && m.PageLocator == nil {
			return bad("citation requires locator")
		}
		u, e := url.Parse(m.SourceUrl)
		if e != nil || (u.Scheme != "https" && u.Scheme != "http") || u.Host == "" {
			return bad("invalid citation URL")
		}
	case *pb.Answer:
		if m.Meta.CorpusId != m.Snapshot.CorpusId {
			return bad("answer corpus mismatch")
		}
		claims := map[string]bool{}
		for _, c := range m.Claims {
			if claims[c.ClaimId] {
				return bad("duplicate claim")
			}
			claims[c.ClaimId] = true
			if e := CheckUTF8Span([]byte(m.Text), c.AnswerTextSpan.StartByte, c.AnswerTextSpan.EndByte); e != nil {
				return e
			}
		}
		for _, c := range m.Citations {
			for _, id := range c.ClaimIds {
				if !claims[id] {
					return bad("citation references unknown claim")
				}
			}
		}
		if m.SemanticStatus == pb.SemanticStatus_SEMANTIC_STATUS_COMPLETE && (len(m.MissingEvidence) > 0 || m.CompletionStatus != pb.CompletionStatus_COMPLETION_STATUS_SUCCEEDED) {
			return bad("incomplete answer marked complete")
		}
	case *pb.AnswerEvent:
		if f := m.GetFinal(); f != nil && (f.RequestId != m.RequestId || f.CompletionStatus != pb.CompletionStatus_COMPLETION_STATUS_SUCCEEDED) {
			return bad("invalid FINAL event")
		}
	case *pb.UpdatePlan:
		seen := map[string]bool{}
		for _, id := range m.ReuseSet {
			seen[id] = true
		}
		for _, id := range m.RecomputeSet {
			if seen[id] {
				return bad("reuse/recompute overlap")
			}
		}
	case *pb.ModelManifest:
		if m.Task == pb.ModelTask_MODEL_TASK_EMBED && m.Dimensions == nil {
			return bad("embedding requires dimensions")
		}
	case *pb.EmbedBatchRequest:
		if m.Model.Task != pb.ModelTask_MODEL_TASK_EMBED {
			return bad("wrong embedding model task")
		}
		ids := map[string]bool{}
		for _, x := range m.Items {
			if ids[x.ItemId] {
				return bad("duplicate batch item ID")
			}
			ids[x.ItemId] = true
		}
	case *pb.RerankBatchRequest:
		if m.Model.Task != pb.ModelTask_MODEL_TASK_RERANK {
			return bad("wrong reranker model task")
		}
		ids := map[string]bool{}
		for _, x := range m.Pairs {
			if ids[x.PairId] {
				return bad("duplicate pair ID")
			}
			ids[x.PairId] = true
		}
	case *pb.GateResult:
		if m.Status == pb.GateStatus_GATE_STATUS_PASS && (m.Denominator == 0 || m.Estimate == nil || len(m.RawArtifacts) == 0) {
			return bad("PASS requires samples, estimate, raw evidence")
		}
	}
	return nil
}

// CheckUTF8Span must be called after the referenced text artifact has been resolved.
func CheckUTF8Span(text []byte, start, end uint64) error {
	if !utf8.Valid(text) || start > end || end > uint64(len(text)) {
		return errors.New("invalid UTF-8 text/span bounds")
	}
	if start < uint64(len(text)) && !utf8.RuneStart(text[start]) || end < uint64(len(text)) && !utf8.RuneStart(text[end]) {
		return errors.New("span splits UTF-8 code point")
	}
	return nil
}

// VerifyEmbeddingResults rejects missing/extra/duplicate IDs and model changes across a batch.
func VerifyEmbeddingResults(req *pb.EmbedBatchRequest, res *pb.EmbedBatchResponse) error {
	return VerifyEmbeddingResultsWithLimits(req, res, DefaultWireLimits)
}

// VerifyEmbeddingResultsWithLimits lets native callers account for every vector scalar
// within their explicit byte/item budget without duplicating correlation rules.
func VerifyEmbeddingResultsWithLimits(req *pb.EmbedBatchRequest, res *pb.EmbedBatchResponse, limits WireLimits) error {
	if err := ValidateWire(req, DefaultWireLimits); err != nil {
		return err
	}
	if err := ValidateWire(res, limits); err != nil {
		return err
	}
	if req.Context.RequestId != res.RequestId || !proto.Equal(req.Model, res.Model) {
		return errors.New("batch response context/model mismatch")
	}
	expected := map[string]bool{}
	for _, x := range req.Items {
		expected[x.ItemId] = true
	}
	for _, x := range res.Results {
		if !expected[x.ItemId] {
			return errors.New("unexpected/duplicate result ID")
		}
		delete(expected, x.ItemId)
		if v := x.GetEmbedding(); v != nil && uint32(len(v.Values)) != req.Model.GetDimensions() {
			return errors.New("embedding result dimension mismatch")
		}
	}
	if len(expected) > 0 {
		return errors.New("batch results missing IDs")
	}
	return nil
}

type AnswerStreamValidator struct {
	RequestID string
	Sequence  uint64
	Terminal  bool
}

func (s *AnswerStreamValidator) Accept(event *pb.AnswerEvent) error {
	if s.Terminal {
		return errors.New("event after terminal")
	}
	if err := ValidateWire(event, DefaultWireLimits); err != nil {
		return err
	}
	if s.RequestID != "" && s.RequestID != event.RequestId || event.StreamSequence != s.Sequence+1 {
		return errors.New("stream correlation/sequence mismatch")
	}
	s.RequestID = event.RequestId
	s.Sequence = event.StreamSequence
	s.Terminal = event.GetFinal() != nil || event.GetError() != nil
	return nil
}
func (s *AnswerStreamValidator) Finish() error {
	if !s.Terminal {
		return errors.New("stream ended without terminal event")
	}
	return nil
}

// EqualPayload compares idempotency fingerprints, not raw non-canonical Protobuf encoding.
func EqualPayload(a, b *pb.ContentHash) bool {
	return a != nil && b != nil && bytes.Equal([]byte(a.Sha256), []byte(b.Sha256))
}

// Infer only explicitly present corpus context; absence is checked by declared required rules.
func wireCorpus(m protoreflect.Message) string {
	if f := m.Descriptor().Fields().ByName("corpus_id"); f != nil {
		return m.Get(f).String()
	}
	for _, name := range []protoreflect.Name{"meta", "context", "batch"} {
		if f := m.Descriptor().Fields().ByName(name); f != nil && f.Kind() == protoreflect.MessageKind && m.Has(f) {
			if c := wireCorpus(m.Get(f).Message()); c != "" {
				return c
			}
		}
	}
	return ""
}
